src/common/Throttle.h

```diff
@@ -330,4 +331,76 @@ class OrderedThrottle {
   uint32_t waiters = 0;
 };
 
+
+class TokenBucketThrottle {
+
+  struct Bucket {
+    CephContext *cct;
+    const std::string name;
+    std::atomic<uint64_t> remain = { 0 }, max = { 0 };
+
+    Bucket(CephContext *cct, const std::string& n, uint64_t m)
+      : cct(cct), name(n),
+	remain(m), max(m)
+    {
+    }
+
+    uint64_t get(uint64_t c);
+    uint64_t put(uint64_t c);
+    void set_max(uint64_t m);
+  };
+
+  struct Blocker {
+    uint64_t tokens_requested;
+    Context *ctx;
+
+    Blocker(uint64_t _tokens_requested, Context* _ctx)
+      : tokens_requested(_tokens_requested), ctx(_ctx) {}
+  };
+
+  CephContext *m_cct;
+  Bucket m_throttle;
+  uint64_t m_avg = 0;
+  SafeTimer *m_timer;
+  Mutex *m_timer_lock;
+  FunctionContext *m_token_ctx;
+  list<Blocker*> m_blockers;
+  Mutex m_lock;
+
+public:
+TokenBucketThrottle(CephContext *cct, uint64_t capacity, uint64_t avg,
+		    SafeTimer *timer, Mutex *timer_lock);
+
+~TokenBucketThrottle();
+
+template <typename T, typename I, void(T::*MF)(int, I*)>
+bool get(uint64_t c, T *handler, I *item) {
+  if (0 == m_throttle.max)
```

Comment:
Nit: unnecessary -- ```Bucket::get``` already handles this case

Reply:
But we can return earlier here to avoid a locking.